File: RemoveRecursionRefactoring/src/ro/pub/cs/diploma/RemoveRecursionInspection.java
package ro.pub.cs.diploma;

import com.intellij.codeInspection.ProblemDescriptor;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiMethod;
import com.intellij.psi.PsiMethodCallExpression;
import com.intellij.psi.PsiModifier;
import com.intellij.psi.PsiParameter;
import com.siyeh.ig.BaseInspection;
import com.siyeh.ig.BaseInspectionVisitor;
import com.siyeh.ig.InspectionGadgetsFix;
import com.siyeh.ig.performance.TailRecursionInspection;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * @see TailRecursionInspection
 */
public class RemoveRecursionInspection extends BaseInspection {
  @Override
  @NotNull
  public String getDisplayName() {
    return RemoveRecursionBundle.message("remove.recursion.display.name");
  }

  @Override
  @NotNull
  protected String buildErrorString(Object... infos) {
    return RemoveRecursionBundle.message("remove.recursion.problem.descriptor");
  }

  @Override
  @Nullable
  protected InspectionGadgetsFix buildFix(Object... infos) {
    final PsiMethod containingMethod = (PsiMethod)infos[0];
    if (!mayBeReplacedByIterativeMethod(containingMethod)) {
      return null;
    }
    return new InspectionGadgetsFix() {
      @Override
      @NotNull
      public String getFamilyName() {
        return RemoveRecursionBundle.message("remove.recursion.replace.quickfix");
      }

      @Override
      public void doFix(Project project, ProblemDescriptor descriptor) {
        final PsiMethod method = Util.INSTANCE.getContainingMethod(descriptor.getPsiElement());
        if (method == null) {
          return;
        }
        IterativeMethodGenerator.getInstance(Util.INSTANCE.getFactory(method), Util.INSTANCE.getStyleManager(method), method, NameManager.getInstance(method))
          .createIterativeBody(13);
      }
    };
  }

  private static boolean mayBeReplacedByIterativeMethod(PsiMethod containingMethod) {
    if (containingMethod.isVarArgs()) {
      return false;
    }
    final PsiParameter[] parameters = containingMethod.getParameterList().getParameters();
    for (final PsiParameter parameter : parameters) {
      if (parameter.hasModifierProperty(PsiModifier.FINAL)) {
        return false;
      }
    }
    return true;
  }

  @Override
  public BaseInspectionVisitor buildVisitor() {
    return new BaseInspectionVisitor() {
      @Override
      public void visitMethodCallExpression(PsiMethodCallExpression expression) {
        super.visitMethodCallExpression(expression);
        final PsiMethod method = Util.INSTANCE.getContainingMethod(expression);
        if (method == null) {
          return;
        }
        if (RecursionUtil.isRecursive(expression, method)) {
          registerMethodCallError(expression, method);
        }
      }
    };
  }
}

File: RemoveRecursionRefactoring/src/ro/pub/cs/diploma/RemoveRecursionInspection.kt
package ro.pub.cs.diploma

import com.intellij.codeInspection.ProblemDescriptor
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiMethod
import com.intellij.psi.PsiMethodCallExpression
import com.intellij.psi.PsiModifier
import com.siyeh.ig.BaseInspection
import com.siyeh.ig.BaseInspectionVisitor
import com.siyeh.ig.InspectionGadgetsFix
import com.siyeh.ig.performance.TailRecursionInspection

/**
 * @see TailRecursionInspection
 */
class RemoveRecursionInspection : BaseInspection() {
  override fun getDisplayName(): String = RemoveRecursionBundle.message("remove.recursion.display.name")

  override fun buildErrorString(vararg infos: Any): String = RemoveRecursionBundle.message("remove.recursion.problem.descriptor")

  override fun buildFix(vararg infos: Any): InspectionGadgetsFix? {
    val containingMethod = infos[0] as PsiMethod
    return if (!mayBeReplacedByIterativeMethod(containingMethod)) {
      null
    }
    else object : InspectionGadgetsFix() {
      override fun getFamilyName(): String = RemoveRecursionBundle.message("remove.recursion.replace.quickfix")

      public override fun doFix(project: Project, descriptor: ProblemDescriptor) {
        val method = descriptor.psiElement.getContainingMethod() ?: return
        IterativeMethodGenerator.getInstance(method.getFactory(), method.getStyleManager(), method, NameManager(method))
            .createIterativeBody(13)
      }
    }
  }

  private fun mayBeReplacedByIterativeMethod(containingMethod: PsiMethod): Boolean {
    if (containingMethod.isVarArgs) {
      return false
    }
    val parameters = containingMethod.parameterList.parameters
    return parameters.none { it.hasModifierProperty(PsiModifier.FINAL) }
  }

  override fun buildVisitor(): BaseInspectionVisitor {
    return object : BaseInspectionVisitor() {
      override fun visitMethodCallExpression(expression: PsiMethodCallExpression) {
        super.visitMethodCallExpression(expression)
        val method = expression.getContainingMethod() ?: return
        if (expression.isRecursiveCallTo(method)) {
          registerMethodCallError(expression, method)
        }
      }
    }
  }
}